File: Week 5 - 03.03.2021 DAA Lab/Optimal_Merge_Pattern.cpp
// Author : Abhishek Sharma
// Optimal Merge Pattern using CPP language
// Design analysis and algorithm lab problems
// difficulty : medium
// score : 10


#include<bits/stdc++.h> 
using namespace std; 
int minComputation(int size, int files[]) 
{  
    priority_queue<int, vector<int>, 
        greater<int>> pq; 

    for(int i = 0; i < size; i++) 
    { 
        pq.push(files[i]); 
    } 
    int count = 0; 
    // Author : Abhishek Sharma
    while(pq.size() > 1) 
    { 
        int first_smallest = pq.top(); 
        pq.pop(); 
        int second_smallest = pq.top(); 
        pq.pop(); 
        
        int temp = first_smallest + second_smallest; 
        count += temp; 
        pq.push(temp); 
    } 
    return count; 
} 

int main() 
{ 
    int n = 6; 
    int files[] = { 5, 3, 2, 7, 9, 13 }; 
    // Author : Abhishek Sharma
    cout << minComputation(n, files); 
    return 0; 
} 

File: Week 5 - 03.03.2021 DAA Lab/Optimal_Merge_Pattern.java
// Author : Abhishek Sharma
// Optimal Merge Pattern Problem using JAVA
// difficulty : medium
// score : 10


import java.util.Scanner; 
import java.util.PriorityQueue; 
// Author : Abhishek Sharma
public class Solution { 

    static int minComputation(int size, int files[]) 
    { 
        // pythonpapa
        PriorityQueue<Integer> pq 
            = new PriorityQueue<>(); 

        for (int i = 0; i < size; i++) { 

            pq.add(files[i]); 
        } 
        // laalbaba
        int count = 0; 

        while (pq.size() > 1) { 

            int temp = pq.poll() + pq.poll(); 

            count += temp; 

            pq.add(temp); 
        } 
        //pythonpapa written JAVA code!
        return count; 
    } 
// driver code by pythonpapa
    public static void main(String[] args) 
    { 


        int size = 6; 

        int files[] = new int[] { 5,3,2,7,9,13 }; 

        System.out.println(minComputation(size, files)); 
    } 
} 
// code contributed by Abhishek Sharma, 2021 @abhisheks008
